main/lsp/test/error_reporter_test.cc

```diff
@@ -0,0 +1,191 @@
+        EXPECT_TRUE(output.empty());
+    }
+}
+
+TEST(ErrorReporterTest, ErrorReporterIgnoresErrorsFromOldEpochs) {
+    auto gs = makeGS();
+    auto cs = makeConfig();
+    ErrorReporter er(cs);
+    vector<unique_ptr<core::Error>> emptyErrorList;
+    vector<unique_ptr<core::Error>> errors;
+    {
+        core::UnfreezeFileTable fileTableAccess(*gs);
+        auto initialEpoch = 0;
+        auto slowPathEpoch = 1;
+        auto latestEpoch = 2;
+        auto file = make_shared<core::File>("foo.rb", "foo", core::File::Type::Normal, initialEpoch);
+        auto fref = gs->enterFile(file);
+        auto outputVector = dynamic_pointer_cast<LSPOutputToVector>(cs->output);
+        errors.emplace_back(
+            make_unique<core::Error>(core::Loc(fref, 0, 0), core::ErrorClass{1, core::StrictLevel::True}, "MyError",
+                                     vector<core::ErrorSection>(), vector<core::AutocorrectSuggestion>(), false));
+
+        // pushDiagnostics is called for foo.rb at epoch 0 with no errors (initial state)
+        er.pushDiagnostics(initialEpoch, fref, emptyErrorList, *gs);
+        // pushDiagnostics is called for foo.rb at epoch 2 with no errors (the fast path preemption)
+        er.pushDiagnostics(latestEpoch, fref, emptyErrorList, *gs);
+        // pushDiagnostics is called for foo.rb at epoch 1 with errors (the slow path)
+        er.pushDiagnostics(slowPathEpoch, fref, errors, *gs);
+
+        // We expect that no errors are reported to VS Code
+        auto output = outputVector->getOutput();
+        EXPECT_TRUE(output.empty());
+    }
+}
+TEST(ErrorReporterTest, filesUpdatedSince) {
+    auto cs = makeConfig();
+    auto gs = makeGS();
+    ErrorReporter er(cs);
+    vector<unique_ptr<core::Error>> errors;
+    vector<unique_ptr<core::Error>> emptyErrorList;
+    {
+        core::UnfreezeFileTable fileTableAccess(*gs);
+        auto epoch = 0;
+        auto requestedEpoch = 3;
+        auto file = make_shared<core::File>("foo.rb", "foo", core::File::Type::Normal, epoch);
+        auto fref = gs->enterFile(file);
+        auto fileWithoutErrors = make_shared<core::File>("bar.rb", "bar", core::File::Type::Normal, epoch);
+        auto frefWithoutErrors = gs->enterFile(fileWithoutErrors);
+        errors.emplace_back(
+            make_unique<core::Error>(core::Loc(fref, 0, 0), core::ErrorClass{1, core::StrictLevel::True}, "MyError",
+                                     vector<core::ErrorSection>(), vector<core::AutocorrectSuggestion>(), false));
+
+        er.pushDiagnostics(epoch, fref, errors, *gs);
+        EXPECT_TRUE(er.filesUpdatedSince(requestedEpoch).empty())
+            << fmt::format("Only returns files with lastReportedEpoch >= sent epoch");
+
+        er.pushDiagnostics(requestedEpoch, fref, errors, *gs);
+        er.pushDiagnostics(requestedEpoch, frefWithoutErrors, emptyErrorList, *gs);
+        EXPECT_EQ(1, er.filesUpdatedSince(requestedEpoch).size()) << fmt::format("Only returns files with errors");
```

Comment:
Can you also check that the return value only contains `fref` and not `frefWithoutErrors`?